### ISSUE TITLE
Overloads with `-> Never` for invalid uses of `contextlib.ContextDecorator`?

### ISSUE THREAD

**Issue by Zac-HD** [python/typeshed#12881]
https://github.com/python/cpython/issues/125862 describes some problems that can occur when applying a `ContextDecorator` to an async function or a generator function.  Is it reasonable to add overloads returning `Never` for such cases, so that typecheckers can warn people?

The code in question _runs_, but it doesn't do anything useful let alone have the expected effect, so I'd love to warn about it sooner than CPython will release a fix.

**Comment by AlexWaygood** [comment]
Yes, when we've tried these kinds of `-> Never` overloads in the past it's almost always caused more trouble than it's worth, unfortunately (see https://github.com/python/typeshed/issues/8566 for an example)

**Comment by Zac-HD** [comment]
Makes sense!  Hopefully we'll get a decision upstream, to at least document it, and can then use `@deprecated` in typeshed.
